BookMyName DNS Authenticator

aaribaud opened the pull request:
Implement an Authenticator which can fulfill a dns-01 challenge using
BookMyName. Applicable only for domains using BookMyName for DNS.

NOTE: BookMyName provides an API for major users (more than 500 domains)
but this Authenticator does not use the API; it uses regular HTTPS.

Testing Done:
 * `tox -e py27`
 * `tox -e lint`
 * `tox -e cover` (100% coverage)
 * Manual testing:
    * Used `certbot certonly --dns-bookmyname -d`, specifying a
      credentials file as a command line argument. Verified that a
      certificate was successfully obtained without user interaction.
    * Used `certbot certonly --dns-bookmyname -d`, without specifying a
      credentials file as a command line argument. Verified that the user
      was prompted and that a certificate was successfully obtained.
    * Used `certbot certonly -d`. Verified that the user was prompted for
      a credentials file after selecting bookmyname interactively and
      that a certificate was successfully obtained.
    * Used `certbot renew --force-renewal`. Verified that certificates
      were renewed without user interaction.
 * Negative testing:
    * Path to non-existent credentials file.
    * Credentials file with unsafe permissions (644).
    * Credentials file missing login or password.
    * Credentials file with blank or incorrect login.
    * Credentials file with blank or incorrect password.
    * Domain name not registered to BookMyName account.

aaribaud commented:
Hey @zjs, should I wait for your review before I look into the tox.cover.sh conflict, or should I go ahead and fix right away?

zjs commented on certbot-dns-bookmyname/setup.py:
```diff
@@ -0,0 +1,69 @@
+from setuptools import find_packages
+
+
+version = '0.25.0.dev0'
```
It looks like this should now be `0.26.0.dev0` (sorry for the delay with this review!)

aaribaud replied:
No prob. I'll update the branch against current head

aaribaud replied:
Done

zjs commented on certbot-dns-bookmyname/certbot_dns_bookmyname/__init__.py:
```diff
@@ -0,0 +1,87 @@
+path to this file for use during renewal, but does not store the file's contents.
+
+.. caution::
+   You should protect these API credentials as theyinclude the password to your
```
nit-pick: "theyinclude" should be "they include"

aaribaud replied:
Will fix

aaribaud replied:
Done

zjs commented on certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py:
```diff
@@ -0,0 +1,330 @@
+        of domains. Each domain anchor in the first column provides the
+        name (in the inner HTML) and the ID (at the end of the href URI)
+
+        WATCH OUT: this has not been tested against cases where not all
```
This scenario concerns me, especially because of the short validity of Let's Encrypt's certificates.

Someone could begin using this plugin with a set of domains, configure a cron job for certbot to renew their certificates automatically, and then register additional domains with BookMyName causing some domains to be pushed onto a second page, at which point some certificates could no longer be renewed.

How hard would it be to handle paging?

aaribaud replied:
The main problem is that I don't have enough BMN domains myself there to be able to tell how multiple domain pages actually look like (or even if they do appear), and to test the plugin against those cases. I'll try and look for someone who has and is willing to help, but I can't make promises.

aaribaud replied:
FYI, I've obtained from the BookMyName support desk a way to really get all domains.

aaribaud replied:
New PR uses a URI which returns all domains in a CSV list (names, not IDs, but IDs are now unnecessary)

zjs commented on certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py:
```diff
@@ -0,0 +1,330 @@
+                .format(registered_domain, short_valid_name, validation, e)
+            )
+
+    def del_txt_record(self, domain, validation_name, validation):
```
This method and `add_txt_record` seems to include a lot of duplicate logic. Can we introduce some private methods to consolidate the common behaviors?

aaribaud replied:
Sure.

aaribaud replied:
Done

zjs commented on certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py:
```diff
@@ -0,0 +1,330 @@
+        This method will log in if needed.
+        """
+        self.login()
+        for registered_domain in self.domains_and_ids:
```
How about something like:

```python
domains = [registered for registered in self.domains_and_ids if domain.endswith(registered)]
if len(domains) == 0:
    raise errors.PluginError(
    // ... no registered domain
    )

if len(domains) > 1:
    raise errors.PluginError(
    // ... too many registered domains ...
    )

return domains[0]
```

I can't think of a case where there _would be_ multiple matching domains, but it seems better to handle that explicitly than just pick the first we encounter.

aaribaud replied:
Ok

aaribaud replied:
Done

zjs commented on certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py:
```diff
@@ -0,0 +1,330 @@
+        This method will log in if needed.
+        """
+        self.login()
+        registered_domain = self.get_registered_domain(domain)
```
Currently, it seems like we invoke `get_registered_domain` twice (once to get the domain, once here to get the ID — and we get both). This seems a little inefficient.

Could we have a single method that just returns the domain and its ID?

aaribaud replied:
Trying a few things, I'll get back to you as soon as I'm done.

aaribaud replied:
OK, I manually checked that in places where I use the id, I can also use the domain, so the id becomes unnecessary.

aaribaud replied:
Domain IDs are not used any more, only domain names.

aaribaud commented:
Hmm, anyone know the reason why a new continuous-integration/appveyor/pr check has appeared and seems stalled?

bmw commented:
It's a test that we added since you submitted the PR so nothing triggered the build. Opening and closing the PR often fixes it, however, GitHub is currently having some issues with their hooks so it may take a while (see their [status page](https://status.github.com/messages)).

Alternatively, merging in `master` will solve the problem.

aaribaud commented:
Well, closing/reopening _did_ something, as the check now has a "Details" link to the AppVeyor jobs queue page. I'll keep an eye on it. Thanks!

aaribaud commented:
Seems like the "Travis CI - Pull Request" check is blocked like the continuous-integration/appveyor was. Closing and re-opening the PR, just in case...

hobbe commented:
Hi @aaribaud , any idea if this plugin will make it to certbot one day?

aaribaud commented:
> Hi @aaribaud , any idea if this plugin will make it to certbot one day?

I did spend some time keeping this patch alive and rebased, but after quite a few months, I moved on to another solution: I now use ALPN challenge method along with wildcard DNS entries, so I don't need DNS method any more. Feel free to take over!

bmw commented:
we're no longer accepting new first party plugins so i'm going to close this

thanks for working on this though. if you'd like to offer this a 3rd party plugin and open a PR advertising it [here](https://eff-certbot.readthedocs.io/en/stable/using.html#third-party-plugins) like many others have done, you are welcome to do so

Neustradamus commented:
Dear @bmw,

Can you add BookMyName feature, it exists since a very long time.

Thanks in advance.

bmw commented:
if you'd like to open a PR to do so i can review it!

here's an example of a recent PR that did this for another plugin https://github.com/certbot/certbot/pull/9879

feel free to @ me in the PR description if you end up creating one

Neustradamus commented:
@bmw: Can you check?
- https://github.com/Joffr3y/certbot-dns-bookmyname
